src/device_imxrt.c

```diff
@@ -0,0 +1,278 @@
+#include "device.h"
+#include "log.h"
+
+#if MG_DEVICE == MG_DEVICE_RT1020 || MG_DEVICE == MG_DEVICE_RT1060
+
+#include <flexspi.h>  /**************************************************************************************************************************/
+// identical for both 1020 and 1060, they have the same controller. This should be part of Mongoose, not read from the example
+
+
+// these are defined in the device CMSIS .h file, so examples will have their own, we should use a different name or #undef
+// hence, the example flexspi.h does not have these macros, that is NXP code carrying their license
+// Mongoose does not include any CMSIS file but these would go into mongoose.h and the user including mongoose.h and the device .h both definitions would clash
+#ifndef FLEXSPI_LUT_OPERAND0
+#define FLEXSPI_LUT_OPERAND0(x) (((uint32_t) (((uint32_t) (x)))) & 0xFFU)
+#define FLEXSPI_LUT_NUM_PADS0(x) \
+  (((uint32_t) (((uint32_t) (x)) << 8U)) & 0x300U)
+#define FLEXSPI_LUT_OPCODE0(x) \
+  (((uint32_t) (((uint32_t) (x)) << 10U)) & 0xFC00U)
+#define FLEXSPI_LUT_OPERAND1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 16U)) & 0xFF0000U)
+#define FLEXSPI_LUT_NUM_PADS1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 24U)) & 0x3000000U)
+#define FLEXSPI_LUT_OPCODE1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 26U)) & 0xFC000000U)
+#endif
+
+#define FLEXSPI_NOR_INSTANCE 0
+
+#if MG_DEVICE == MG_DEVICE_RT1020
+typedef struct {
+  uint32_t version;
+  int (*init)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*program)(uint32_t instance, flexspi_nor_config_t *config,
+                      uint32_t dst_addr, const uint32_t *src);
+  uint32_t reserved;
+  int (*erase)(uint32_t instance, flexspi_nor_config_t *config,
+                    uint32_t start, uint32_t lengthInBytes);
+  uint32_t reserved2;
+  int (*update_lut)(uint32_t instance, uint32_t seqIndex,
+                         const uint32_t *lutBase, uint32_t seqNumber);
+  int (*xfer)(uint32_t instance, char *xfer);
+  void (*clear_cache)(uint32_t instance);
+} flexspi_nor_driver_interface_t;
+#elif MG_DEVICE == MG_DEVICE_RT1060
+typedef struct {
+  uint32_t version;
+  int (*init)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*program)(uint32_t instance, flexspi_nor_config_t *config,
+                      uint32_t dst_addr, const uint32_t *src);
+  int (*erase_all)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*erase)(uint32_t instance, flexspi_nor_config_t *config,
+                    uint32_t start, uint32_t lengthInBytes);
+  int (*read)(uint32_t instance, flexspi_nor_config_t *config,
+                   uint32_t *dst, uint32_t addr, uint32_t lengthInBytes);
```

Comment:
These will also probably clash with the xip.c code in Xpresso, we should come up with our own naming or adding the usual prefix combo

Reply:
Yes.
We should redefine our own structure.
The same way we do e.g. for ST Ethernet driver, or ST H5/H7 devices for flashing - we do not rely on any SDK sources, but define addresses/structures ourselves.